src/data/nat/factorization.lean

```diff
@@ -0,0 +1,134 @@
+/-- For nonzero `a` and `b`, the power of `p` in `a * b` is the sum of the powers in `a` and `b` -/
+@[simp] lemma factorization_mul {a b : ℕ} (ha : a ≠ 0) (hb : b ≠ 0) :
+  (a * b).factorization = a.factorization + b.factorization :=
+by { ext p, simp only [add_apply, factorization_eq_count,
+  count_factors_mul_of_pos (zero_lt_iff.mpr ha) (zero_lt_iff.mpr hb)] }
+
+/-- For any `p`, the power of `p` in `n^k` is `k` times the power in `n` -/
+lemma factorization_pow {n k : ℕ} :
+  factorization (n^k) = k • n.factorization :=
+begin
+  ext p,
+  simp only [algebra.id.smul_eq_mul, finsupp.coe_smul, pi.smul_apply],
+  simp only [factorization_eq_count, factors_count_pow],
+end
+
+/-- The only prime factor of prime `p` is `p` itself, with multiplicity `1` -/
+@[simp] lemma prime.factorization {p : ℕ} (hp : prime p) :
+  p.factorization = single p 1 :=
+begin
+  ext q,
+  rw [factorization_eq_count, factors_prime hp],
+  by_cases hqp : q = p,
+  { rw hqp, simp },
+  { rw finsupp.single_eq_of_ne (ne.symm hqp),
+    exact count_eq_zero_of_not_mem ((not_iff_not_of_iff list.mem_singleton).mpr hqp) },
+end
+
+/-- For prime `p` the only prime factor of `p^k` is `p` with multiplicity `k` -/
+@[simp] lemma prime.factorization_pow {p k : ℕ} (hp : prime p) :
+  factorization (p^k) = single p k :=
+by simp [factorization_pow, prime.factorization hp]
+
+/-! ### Factorizations of pairs of coprime numbers -/
+
+/-- The prime factorizations of coprime `a` and `b` are disjoint -/
+lemma factorization_disjoint_of_coprime {a b : ℕ} (hab : coprime a b) :
+  disjoint a.factorization.support b.factorization.support :=
+by simpa only [support_factorization]
+  using disjoint_to_finset_iff_disjoint.mpr (coprime_factors_disjoint hab)
+
+/-- For coprime `a` and `b`, the power of `p` in `a * b` is the sum of the powers in `a` and `b` -/
+lemma factorization_mul_of_coprime {a b : ℕ} (hab : coprime a b) :
+  (a * b).factorization = a.factorization + b.factorization :=
+begin
+  ext q,
+  simp only [finsupp.coe_add, pi.add_apply, factorization_eq_count],
+  simp only [count_factors_mul_of_coprime hab],
+end
+
+/-- For coprime `a` and `b` the prime factorization `a * b` is the union of those of `a` and `b` -/
+lemma factorization_mul_support_of_coprime {a b : ℕ} (hab : coprime a b) :
+  (a * b).factorization.support =
+    a.factorization.support ∪ b.factorization.support :=
+begin
+  rw factorization_mul_of_coprime hab,
+  exact support_add_eq (factorization_disjoint_of_coprime hab),
+end
+
+/-- For any multiplicative function `f` with `f 1 = 1` and any `n > 0`,
+we can evaluate `f n` by evaluating `f` at `p ^ k` over the factorization of `n` -/
+lemma multiplicative_factorization {n : ℕ} {β : Type*} [comm_monoid β] {f : ℕ → β}
+  (hn : 0 < n) (h_mult : ∀ x y : ℕ, coprime x y → f(x * y) = f x * f y) (hf : f 1 = 1) :
+f n = n.factorization.prod (λ p k, f(p ^ k)) :=
+begin
+  apply @nat.rec_on_pos_prime_coprime
+    (λ n, (0 < n) → (f n = n.factorization.prod (λ p k, f(p ^ k)))),
```

Comment:
```suggestion
lemma multiplicative_factorization {β : Type*} [comm_monoid β] {f : ℕ → β}
  (h_mult : ∀ x y : ℕ, coprime x y → f(x * y) = f x * f y) (hf : f 1 = 1) :
∀ {n : ℕ}, (0 < n) → f n = n.factorization.prod (λ p k, f(p ^ k)) :=
begin
  apply' nat.rec_on_pos_prime_coprime,
```

this gets rid of the `exact hn` and also the ugly `@`

Reply:
maybe it's also worth making a version that removes the 0 < n condition in exchange for `f 0 = 1`

Reply:
Ah, much better, thanks!

Reply:
Thanks for the suggestion, I've added that alternative version as `multiplicative_factorization'`